trin-types/src/content_item.rs

```diff
@@ -0,0 +1,158 @@
+use crate::execution::accumulator::EpochAccumulator;
+use crate::execution::block_body::BlockBody;
+use crate::execution::header::HeaderWithProof;
+use crate::execution::receipts::Receipts;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use ssz::{Decode, Encode};
+use trin_utils::bytes::{hex_decode, hex_encode};
+
+/// An encodable portal network content item.
+pub trait ContentItem: Sized {
+    /// Encodes the content item, appending the encoded bytes to `buf`.
+    fn encode(&self) -> Vec<u8>;
+    /// Decodes `buf` into a content item.
+    fn decode(buf: &[u8]) -> anyhow::Result<Self>;
+}
+
+/// The length of the Merkle proof for the inclusion of a block header in a particular epoch
+/// accumulator.
+pub const EPOCH_ACC_PROOF_LEN: usize = 15;
+
+/// A Portal History content item.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[allow(clippy::large_enum_variant)]
+pub enum HistoryContentItem {
+    BlockHeaderWithProof(HeaderWithProof),
+    BlockBody(BlockBody),
+    Receipts(Receipts),
+    EpochAccumulator(EpochAccumulator),
+    /// A placeholder for data that could not be interpreted as any of the valid content types.
+    Unknown(String),
+}
+
+impl ContentItem for HistoryContentItem {
+    fn encode(&self) -> Vec<u8> {
+        match self {
+            Self::BlockHeaderWithProof(item) => item.as_ssz_bytes(),
+            Self::BlockBody(item) => item.as_ssz_bytes(),
+            Self::Receipts(item) => item.as_ssz_bytes(),
+            Self::EpochAccumulator(item) => item.as_ssz_bytes(),
+            Self::Unknown(item) => hex_decode(item).unwrap_or_default(),
+        }
+    }
+
+    fn decode(buf: &[u8]) -> anyhow::Result<Self> {
+        if let Ok(item) = HeaderWithProof::from_ssz_bytes(buf) {
+            return Ok(Self::BlockHeaderWithProof(item));
+        }
+
+        if let Ok(item) = BlockBody::from_ssz_bytes(buf) {
+            return Ok(Self::BlockBody(item));
+        }
+
+        if let Ok(item) = Receipts::from_ssz_bytes(buf) {
+            return Ok(Self::Receipts(item));
+        }
+
+        if let Ok(item) = EpochAccumulator::from_ssz_bytes(buf) {
+            return Ok(Self::EpochAccumulator(item));
+        }
+
+        Ok(Self::Unknown(hex_encode(buf)))
+    }
+}
+
+impl Serialize for HistoryContentItem {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let encoded = match self {
+            Self::BlockHeaderWithProof(item) => item.as_ssz_bytes(),
+            Self::BlockBody(item) => item.as_ssz_bytes(),
+            Self::Receipts(item) => item.as_ssz_bytes(),
+            Self::EpochAccumulator(item) => item.as_ssz_bytes(),
+            Self::Unknown(item) => {
+                if item.is_empty() {
+                    return serializer.serialize_str("0x0");
+                }
+                item.clone().into_bytes()
+            }
```

Comment:
So yeah, I'd be curious what goes wrong if we return an `Err` here, instead? I'm assuming that introduces challenges somewhere that I can't see by reading this code block.

Even if it's not changed during this merge, a code comment and some info in the tracking issue about what happens when attempting this change would be valuable.

Reply:
I believe this is handling the case of a failed attempt to lookup a content key in local storage. In which case we're supposed to return `0x0`. But, iiuc, this will be handled in #612